app/app/Http/Controllers/PasswordRecoveryController.php

```diff
@@ -32,8 +33,49 @@ public function sendEmail(Request $request)
 		if ( !$matching_user ) {
 			return view('pages.password_recovery.unmatched_email');
 		}
-		
-		mail($email, 'Password Recovery', 'Your password has been reset.');
+		BaseUser::sendPasswordRecoveryEmail($matching_user);
 		return view('pages.password_recovery.email_sent');
 	}
-}
\ No newline at end of file
+
+	public function passworkRecover($user_email, $password_recovery_token) {
+		$matching_user = User::where('email', '=', $user_email)->first();
+
+		if ( !$matching_user ) {
+			return view('pages.password_recovery.unmatched_email');
+		}
+
+		if ( $matching_user->password_recovery_token == $password_recovery_token ) {
+			Session::put('PasswordRecovery',$matching_user->email);
+			return view('pages.password_recovery.reset_password');
+		}
+
+		return view('pages.password_recovery.unmatched_token');
+	}
+	public function resetPassword(Request $request) {
+		if ( Session::has('PasswordRecovery') ) {
+			$validation_rules = array(
+				'new_password'         => 'required',
+				'password_confirm'     => 'required|same:new_password'
+			);
+			$validator = Validator::make(Input::all(), $validation_rules);
+			$failing = $validator->fails();
+			if ($failing)
+			{
+				return Redirect::back()->withErrors($validator)->withInput();
+			}
+			else
+			{
+				$user = User::where('email', '=', Session::get('PasswordRecovery'))->first();
+				$user->password_hash = User::generateSaltedHash($request->input('new_password'));
+				$user->save();
+				Session::forget('PasswordRecovery');
```

Comment:
What is the PasswordRecovery session value used for?  What harm would come if all of its references were removed along with the if-statement here?

It seems like if anyone clicks the emailed link, it shouldn't matter if the session variable is set or not.  It should provide a form with the password token in a &lt;input type="hidden".